clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetadata.java

```diff
@@ -72,16 +157,68 @@ public synchronized boolean containsTopic(String topic) {
 
     @Override
     public synchronized boolean retainTopic(String topic, boolean isInternal, long nowMs) {
-        Long expireMs = topics.get(topic);
-        if (expireMs == null) {
-            return false;
-        } else if (expireMs <= nowMs) {
-            log.debug("Removing unused topic {} from the metadata list, expiryMs {} now {}", topic, expireMs, nowMs);
-            topics.remove(topic);
+        TopicInfo topicInfo = topics.get(topic);
+        if (topicInfo == null) {
             return false;
-        } else {
+        }
+
+        long expireMs = topicInfo.lastAccessTime() + TOPIC_EXPIRY_MS;
+        if (expireMs > nowMs) {
             return true;
         }
+
+        log.debug("Removing unused topic {} from the metadata list, expireMs {}, now {}", topic, expireMs, nowMs);
+        refresh.remove(topicInfo);
+        topics.remove(topic);
+        return false;
+    }
+
+    @Override
+    public synchronized void topicMetadataUpdated(String topic, long nowMs) {
+        TopicInfo topicInfo = topics.get(topic);
+        if (topicInfo == null) {
+            return;
+        }
+
+        refresh.remove(topicInfo);
+        topicInfo.updateLastRefreshTime(nowMs);
+        refresh.add(topicInfo);
+    }
+
+    @Override
+    public synchronized long computeTimeToNextUpdate(long nowMs) {
+        long nextUpdateMs = super.computeTimeToNextUpdate(nowMs);
+        if (nextUpdateMs == 0) {
```

Comment:
should this `if (nextUpdateMs == 0) ` be `if (nextUpdateMs <= 0) {`

Reply:
This would be safer, yes. Done.